crates/analyzer/src/traversal/expressions.rs

```diff
@@ -873,15 +879,65 @@ fn check_for_call_to_init_fn(
     }
 }
 
+fn check_for_self_param(
+    scope: &mut BlockScope,
+    use_span: Span,
+) -> Result<(), FatalError> {
+    if scope.root.function.signature(scope.db()).self_decl == SelfDecl::None {
+        scope.fancy_error(
+            "`self` is not defined",
+            vec![Label::primary(use_span, "undefined value")],
+            vec![
+                "add `self` to the scope by including it in the function signature".to_string(),
+                "Example: `fn foo(self, bar: bool)`".to_string()
+            ],
+        );
+        Err(FatalError::new())
+    } else {
+        Ok(())
+    }
+}
+
 fn expr_call_self_attribute(
+    scope: &mut BlockScope,
+    func_name: &str,
+    name_span: Span,
+    self_span: Span,
+    args: &Node<Vec<Node<fe::CallArg>>>,
+) -> Result<ExpressionAttributes, FatalError> {
+    check_for_call_to_init_fn(scope, func_name, name_span)?;
+    check_for_self_param(scope, self_span)?;
+
+    if let Some(func) = scope.root.self_contract_function(func_name) {
+        let sig = func.signature(scope.root.db);
+        validate_arg_count(scope, func_name, name_span, args, sig.params.len());
+        validate_arg_types(scope, func_name, args, &sig.params)?;
+
+        let return_type = sig.return_type.clone()?;
+        let return_location = Location::assign_location(&return_type);
+        Ok(ExpressionAttributes::new(
+            return_type.into(),
+            return_location,
+        ))
+    } else {
+        scope.fancy_error(
+            &format!("no function `{}` exists this contract", func_name),
+            vec![Label::primary(name_span, "undefined function")],
+            vec![],
+        );
+        Err(FatalError::new())
+    }
+}
+
+fn expr_call_pure(
     scope: &mut BlockScope,
     func_name: &str,
     name_span: Span,
     args: &Node<Vec<Node<fe::CallArg>>>,
 ) -> Result<ExpressionAttributes, FatalError> {
     check_for_call_to_init_fn(scope, func_name, name_span)?;
 
-    if let Some(func) = scope.root.contract_function(func_name) {
+    if let Some(func) = scope.root.pure_contract_function(func_name) {
```

Comment:
the error for attempting to call a `self` function as pure should be more helpful